### PR TITLE
Bug 1875528: Allow reconfiguration of Windows machines after partial configuration failure

### PR DESCRIPTION
**Pull request by sebsoto**
This PR includes multiple commits to fix https://bugzilla.redhat.com/show_bug.cgi?id=1875528 and the symptoms related to it.

### CODE REVIEW

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 1]
```diff
@@ -175,9 +176,35 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
...
+			// NodeRef missing. Requeue and hope it is created. It never being created indicates an issue with the
+			// machine api operator
+			return reconcile.Result{}, fmt.Errorf("ready Windows machine %s missing NodeRef", machine.GetName())
+		}
+
+		node := &core.Node{}
+		err := r.client.Get(context.TODO(), kubeTypes.NamespacedName{
+			Namespace: machine.Status.NodeRef.Namespace,
+			Name:      machine.Status.NodeRef.Name,
+		}, node)
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:197]: This needs to be formatted better.

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 2]
```diff
@@ -175,9 +176,35 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
 	}
 	// provisionedPhase is the status of the machine when it is in the `Provisioned` state
 	provisionedPhase := "Provisioned"
-	if machine.Status.Phase == nil || *machine.Status.Phase != provisionedPhase {
-		// Phase can be nil and should be ignored by WMCO
-		// If the Machine is not in provisioned state, WMCO shouldn't care about it
+	// readyPhase is the status of the machine when it is in the `Running` state, indicating that it is configured into a node
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:179]: Please add an explanation for this chunk of code to your commit message. The commit message seems to imply just changing where the version annotation is placed is sufficient.

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 3]
```diff
@@ -175,9 +176,35 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
 	}
 	// provisionedPhase is the status of the machine when it is in the `Provisioned` state
 	provisionedPhase := "Provisioned"
-	if machine.Status.Phase == nil || *machine.Status.Phase != provisionedPhase {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:178]: Why does `*machine.Status.Phase != provisionedPhase` need to be split out into its own `else if` given you are doing the same action even when `*machine.Status.Phase` is nil.

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 4]
```diff
@@ -175,9 +176,35 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
...
+	// readyPhase is the status of the machine when it is in the `Running` state, indicating that it is configured into a node
+	runningPhase := "Running"
+	if machine.Status.Phase == nil {
+		// Phase is nil and should be ignored by WMCO until phase is set
+		return reconcile.Result{}, nil
+	} else if *machine.Status.Phase == runningPhase {
+		// Machine has been configured into a node, we need to ensure that the version annotation exists. If it doesn't
+		// the machine was not fully configured and needs to be configured properly.
+		if machine.Status.NodeRef == nil {
+			// NodeRef missing. Requeue and hope it is created. It never being created indicates an issue with the
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:188]: Or do we delete the machine? But we could be running into a race condition here.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for the quick turn around on this, @sebsoto. Please address my comments.

File: pkg/controller/windowsmachine/windowsmachine_controller.go (hunk 3 above)
Comment (aravindhp) [reply pkg/controller/windowsmachine/windowsmachine_controller.go:178]: Ignore this comment. I figured it out.

File: pkg/controller/windowsmachine/windowsmachine_controller.go [hunk 5]
```diff
@@ -175,9 +176,35 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
...
+		node := &core.Node{}
+		err := r.client.Get(context.TODO(), kubeTypes.NamespacedName{
+			Namespace: machine.Status.NodeRef.Namespace,
+			Name:      machine.Status.NodeRef.Name,
+		}, node)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrapf(err, "could not get node associated with machine %s", machine.GetName())
+		}
+
+		if _, present := node.Annotations[nodeconfig.VersionAnnotation]; present {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windowsmachine_controller.go:202]: Can't this be checked in the predicate? Though you would need to check if it is the expected version there.

Comment (sebsoto) [reply pkg/controller/windowsmachine/windowsmachine_controller.go:200]: That can be checked in @VaishnaviHire PR with the introduction of the node watch

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go [hunk 6]
```diff
@@ -195,10 +195,12 @@ func (nc *nodeConfig) setNode() error {
 		nodes, err := nc.k8sclientset.CoreV1().Nodes().List(context.TODO(),
 			metav1.ListOptions{LabelSelector: WindowsOSLabel})
 		if err != nil {
-			return false, errors.Wrap(err, "could not get list of nodes")
+			log.V(1).Info("node object listing failed for", "VM ID", nc.ID(), "error", err)
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:199]: `Machine ID`. Applies everywhere.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 7]
```diff
@@ -287,6 +287,19 @@ func (vm *windows) transferFiles() error {
...
 	}
 	for src, dest := range srcDestPairs {
+		// Assumption: Ignore transferring files, if they are already present. This is with assumption that Windows VMs
+		// configuration is immutable for a given version of operator. To make it clear, if we want a newer version
+		// of file that we want to transfer, WMCO team should cut a newer version of operator.
+		// The files are tightly coupled with the operator
+		// TODO: Remove this when we do in place upgrades
+		out, err := vm.Run("Test-Path "+dest+"\\"+filepath.Base(src), true)
+		if err != nil {
+			return errors.Wrapf(err, "error while checking the file %s", dest+"\\"+filepath.Base(src))
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/windows/windows.go:297]: `...checking if file exists...`

**Review by aravindhp — APPROVED** [review]
Thanks for working on this, @sebsoto 

/approve

with the caveat that address the minor comments before merging.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 8]
```diff
@@ -328,6 +348,25 @@ func (vm *windows) createService(svc service) error {
...
+// stopService stops the service that was already running
+func (vm *windows) stopService(svc service) error {
+	out, err := vm.Run("sc.exe stop "+svc.Name(), false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to start service with output: %s", out)
+	}
+	return nil
+}
+
+// IsRunning gets the status of given service
```
Comment (mansikulkarni96) [inline pkg/controller/windowsmachine/windows/windows.go:360]: doesn't need to be exported isRunning()

File: pkg/controller/windowsmachine/windows/windows.go [hunk 9]
```diff
@@ -328,6 +348,25 @@ func (vm *windows) createService(svc service) error {
...
+func (vm *windows) IsRunning(svc service) bool {
+	out, err := vm.Run("sc.exe query "+svc.Name(), false)
+	if err != nil {
+		log.Error(err, "failed to start service with output: %s", out)
```
Comment (mansikulkarni96) [inline pkg/controller/windowsmachine/windows/windows.go:364]: failed to *query service

**Review by mansikulkarni96 — COMMENTED** [review]
@sebsoto thanks for working on this. LGTM except for some changes in the commit picked up from Ravi.

File: pkg/controller/windowsmachine/windows/windows.go [hunk 10]
```diff
@@ -328,6 +348,25 @@ func (vm *windows) createService(svc service) error {
...
+		return errors.Wrapf(err, "failed to stop service with output: %s", out)
+	}
+	return nil
+}
+
+// isRunning gets the status of given service
+func (vm *windows) isRunning(svc service) bool {
+	out, err := vm.Run("sc.exe query "+svc.Name(), false)
+	if err != nil {
+		log.Error(err, "failed to query service with output: %s", out)
```
Comment (mansikulkarni96) [inline pkg/controller/windowsmachine/windows/windows.go:364]: this should be expressed in key,value pairs

**Comment by mansikulkarni96** [comment]
LGTM

**Comment by ravisantoshgudimetla** [comment]
/lgtm

Tank you for working on this @sebsoto

**Pull request merged**